[MM-52941] Adds from: suggestion when typing @ in search

harshal2030 opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does, as well as QA test steps (if applicable and if not already added to the Jira ticket).
-->
This pull request allows users to search using '@' handle and 'from:' keyword both
#### Ticket Link
<!--
If applicable, please include both or either of the following links:

Fixes https://github.com/mattermost/mattermost/issues/XXX
Jira https://mattermost.atlassian.net/browse/MM-XXX
-->
Fixes https://github.com/mattermost/mattermost/issues/23863
<!--
If the PR includes UI changes, include screenshots/GIFs.

For an easier comparison of UI changes a table (template below) can be used.

|  before  |  after  |
|----|----|
| <insert before screenshot here> | <insert after screenshot here> |

-->

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates).
* API additions—new endpoint, new response fields, or newly accepted request parameters.
* Database changes (any).
* Schema migration changes. Use the [Schema Migration Template](https://docs.google.com/document/d/18lD7N32oyMtYjFrJKwsNv8yn6Fe5QtF-eMm8nn0O8tk/edit?usp=sharing) as a starting point to capture these details as release notes. 
* Websocket additions or changes.
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating).
* New features and improvements, including behavioral changes, UI changes, and CLI changes.
* Bug fixes and fixes of previous known issues.
* Deprecation warnings, breaking changes, or compatibility notes.

If no release notes are required, write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note
NONE
```

mattermost-build commented:
Hello @harshal2030,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

sbishel reviewed (changes requested):
Code looks good. If possible, can you add some unit tests for this.

crspeller commented on webapp/channels/src/components/suggestion/search_user_provider.tsx:
```diff
@@ -70,7 +70,7 @@ export default class SearchUserProvider extends Provider {
 
     handlePretextChanged(pretext: string, resultsCallback: ResultsCallback<UserProfile>) {
-        const captured = (/\bfrom:\s*(\S*)$/i).exec(pretext.toLowerCase());
+        const captured = (/\bfrom:\s*(\S*)$/i).exec(pretext.replace('@', 'from:').toLowerCase());
```
Can you explain what this is supposed to do?

harshal2030 replied:
when @ is typed, I replace it from 'from:' so to invoke function to show suggestion

crspeller reviewed (changes requested):
Not sure this is the right approach to this. There are legitimate uses for searching for `@somthing` and not getting posts from that person. For instance I might search `@myself` to find places I have been mentioned. 
I think that's what the recent mentions button does. This might break that.

harshal2030 commented:
@sbishel I've added tests according to implementation. @crspeller I've also changed the location to populate it, I've test with recent mention button. both works fine. Please review this

johndavidlugtu commented:
That's a good callout -- I don't think we'd want to lose that capability.

A suggestion is for us to suggest to the user to add `from:` the moment they start their search with @ as opposed to auto-adding `from:`.

The behaviour would be similar to how starting to type "f" shows the different suggested options
<img width="401" alt="Screenshot 2023-08-17 at 6 36 59 PM" src="https://github.com/mattermost/mattermost/assets/101303489/ae62b681-6e21-4fd2-b05d-c49bc6da91a4">

cc: @sbishel @crspeller

harshal2030 commented:
I think this approach would be better given the case.
@sbishel @crspeller should I work on this implementation or it needs more discussion.

sbishel commented:
@johndavidlugtu - Is the idea that If the user didn't want to do a from search but a content search for "@user", they would then delete "from:" before searching?

harshal2030 commented:
@sbishel @crspeller done with the new implementation, please review and let me know if anything needs change

sbishel commented on webapp/channels/src/components/search/search.tsx:
```diff
@@ -54,11 +54,15 @@ const determineVisibleSearchHintOptions = (searchTerms: string, searchType: Sear
     if (shouldShowHintOptions) {
         try {
             newVisibleSearchHintOptions = options.filter((option) => {
+                if (pretext.startsWith('@') && option.searchTerm === 'From:') {
```
See above `.startsWith()` vs `===`.

sbishel reviewed (changes requested):
That works much better, but I don't think we want to display hints if user continues to type after `@`. 
Also, different formatting would improve readability.

thanks.

sbishel commented on webapp/channels/src/components/search/search.tsx:
```diff
@@ -54,11 +54,15 @@ const determineVisibleSearchHintOptions = (searchTerms: string, searchType: Sear
     const penultimatePretext = pretextArray[pretextArray.length - 2];
 
-    const shouldShowHintOptions = penultimatePretext ? !options.some(({searchTerm}) => penultimatePretext.toLowerCase().endsWith(searchTerm.toLowerCase())) : !options.some(({searchTerm}) => searchTerms.toLowerCase().endsWith(searchTerm.toLowerCase()));
+    const shouldShowHintOptions = penultimatePretext ? !options.some(({searchTerm}) => penultimatePretext.toLowerCase().endsWith(searchTerm.toLowerCase())) || penultimatePretext.startsWith('@') : !options.some(({searchTerm}) => searchTerms.toLowerCase().endsWith(searchTerm.toLowerCase())) || searchTerms.startsWith('@');
```
I find the ternary operator here hard to read and lint won't let you split on multiple lines. I suggest moving to an if/then for readability. Similar to...

```suggestion
   let shouldShowHintOptions;
    if (penultimatePretext) {
        shouldShowHintOptions = !(options.some(({searchTerm}) => penultimatePretext.toLowerCase().endsWith(searchTerm.toLowerCase())) && penultimatePretext !== '@');
    } else {
        shouldShowHintOptions = !options.some(({searchTerm}) => searchTerms.toLowerCase().endsWith(searchTerm.toLowerCase())) || searchTerms === '@';
    }
```

From testing, I think `searchTerms.startsWith('@')` should be changed to `searchTerms === '@'`, so that the hint only displays with `@` and not `@anything`.

harshal2030 commented:
@sbishel fixed based on your comment. Please review

johndavidlugtu commented:
Do we have a test server and/or a demo vid?

sbishel commented:
@johndavidlugtu - You can create by adding the above label...

johndavidlugtu commented:
/update-branch

harshal2030 commented:
@sbishel @crspeller We can merge it now?

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

hanzei commented:
/update-branch

hanzei commented:
@crspeller @sbishel @johndavidlugtu Can we merge this PR?
